Alter CCE interface managers for new changes to interpolation

kidder opened the pull request:
## Proposed changes

This is a rebased version of #3422, dropping the commit [Use evolution_less_equal for receive_local_time_stepping](https://github.com/sxs-collaboration/spectre/pull/3422/commits/d9eb48d055990b8e874804937dc8de6e8f411a0a), while also making sure the individual commits compile

This makes a number of changes to the interface for the GH system sending data to CCE based on the newer design considerations from the availability of dense interpolation:
- The interface managers no longer use the time derivatives of the GH variables, as they won't be as easily available from dense output
- The interface managers are no longer factory-creatable; the `GhLockstep` interface manager is the only available method during self-start, and the `GhLocalTimeStepping` interface manager is the only available method during the evolution
- Handle self-start separately using an action that manually calls the appropriate interpolation event, and take the `TimeStepId` as the relevant temporal id for those utilities intended for self start, and `double` for all other cases.


### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
Any usage of the `GhInterfaceManagers` will likely need to be reworked to account for the new interfaces.
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

moxcodes committed:
Reorder `Register` phase to before self-start

Self start with GH-CCE requires the communication via the interpolator, which
requires registration.
It should be fine to perform registration before self-start, as no current
registration process assumes anything about the state of the `TimeStepId`.

moxcodes committed:
Use self-start specific interface manager during self-start

Also, alter GhInterfaceManagers to fit dense interpolation

- dense interpolation sends `double`s for times rather than TimeStepIds
- dense interpolation will not send time derivatives of the evolved
  quantities, so we must perform the CCE interpolation without them.
- remove factory-creation from GhInterfaceManagers; both versions of
  interface manager will still be needed, and need to maintain a common
  interface, but they should not be selectable at runtime; one will be
  used only during self-start, and one only during the main evolution.

moxcodes committed:
Remove InterpolationStrategies

They will not be needed anymore, as CCE will not support any method of data
transfer apart from dense interpolation.

moxcodes committed:
Add InterpolateDuringSelfStart specialized action for CCE-GH

Interpolation for self start cannot follow the standard events and dense
triggers control flow, so needs a specialized action to provide CCE with
boundary data during self start.

moxcodes committed:
Don't send time derivatives for CCE interpolation

moxcodes committed:
Add pup function to circumvent charm bug for map with comparator

Adds a function `pup_override` to use for `std::map`s with custom comparators,
because currently a charm bug causes an invalid `reserve` call for such
containers.
The function is intentionally chosen to not directly interfere with the charm
pup override so that the charm version of the function can be easily used or
tested once the bug is fixed.

kidder commented:
The main changes from rebasing were from interactions with #3896 and #3828  so @knelli2 and @markscheel I'd appreciate you taking a quick glance at the parts of this PR related to those PRs.  @wthrowe, as you reviewed the original, I'd appreciate you taking a quick glance as well.

knelli2 commented on src/Time/TimeAndPrevious.hpp:
```diff
@@ -0,0 +1,73 @@
+ * events used in `EventsAndDenseTriggers`, which has the ability to provide the
+ * previous trigger time for verifying sequencing of communication.
+ */
+struct TimeAndPrevious {
```
This `TimeAndPrevious` struct is identical to `LinkedMesageId<double>` in `DataStructures`. The only thing you need to add to LinkedMessageId is the custom comparator.

knelli2 commented on tests/Unit/Evolution/Systems/Cce/Actions/Test_InterpolateDuringSelfStart.cpp:
```diff
@@ -0,0 +1,153 @@
+  }
+};
+
+struct MockComputeTargetPoints
+    : tt::ConformsTo<intrp::protocols::ComputeTargetPoints> {
```
`#include "ParallelAlgorithms/Interpolation/Protocols/ComputeTargetPoints.hpp`

Similarly for the other protocols.

knelli2 commented on tests/Unit/Evolution/Systems/Cce/Actions/Test_InterpolateDuringSelfStart.cpp:
```diff
@@ -0,0 +1,153 @@
+    ActionTesting::set_phase(make_not_null(&runner), metavars::Phase::Testing);
+
+    // Call the action on all the elements.
+    for (const auto& element_id : element_ids) {
+      ActionTesting::next_action<elem_component>(make_not_null(&runner),
+                                                 element_id);
+    }
```
Are there any `CHECK`s that the action did what it was supposed to or is just running and not crashing a success?

kidder replied:
there are checks being done by the function `InterpolateOnElementTestHelpers::test_interpolate_on_element` called in `run_test`

knelli2 commented on src/ApparentHorizons/ComputeHorizonVolumeQuantities.hpp:
```diff
@@ -68,12 +68,6 @@ struct ComputeHorizonVolumeQuantities
       const InverseJacobian<DataVector, 3, Frame::ElementLogical, TargetFrame>&
           inverse_jacobian);
 
-  using allowed_src_tags =
-      tmpl::list<gr::Tags::SpacetimeMetric<3, Frame::Inertial>,
-                 GeneralizedHarmonic::Tags::Pi<3, Frame::Inertial>,
-                 GeneralizedHarmonic::Tags::Phi<3, Frame::Inertial>,
-                 Tags::deriv<GeneralizedHarmonic::Tags::Phi<3, Frame::Inertial>,
-                             tmpl::size_t<3>, Frame::Inertial>>;
```
I'm a bit confused by this, but that's probably because I don't know CCE that well. What tags are you using in CCE that aren't in this list? From a quick glance it seems like CCE only needs the metric, pi, and phi (as you removed the time derivatives).

wthrowe commented on src/Time/Tags.hpp:
```diff
@@ -22,9 +22,18 @@
 #include "Utilities/TMPL.hpp"
 
+/// \cond
+namespace evolution {
+namespace Tags {
+struct PreviousTriggerTime;
+}  // namespace Tags
+}  // namespace evolution
```
Shouldn't clang-tidy have complained about namespace nesting here?  I don't care if the code is changed, but I'm worried CI might not be working.

kidder replied:
As far as I can tell CI only runs clang-tidy on changed cpp files (see `tools/ClangTidyHash.sh`).  But even if I add the hpp files to the list of modified files, only the cpp files are used...

wthrowe commented on src/Evolution/Systems/Cce/Actions/CMakeLists.txt:
```diff
@@ -20,6 +20,7 @@ spectre_target_headers(
   InitializeFirstHypersurface.hpp
   InitializeWorldtubeBoundary.hpp
+  InterpolateDuringSelfStart.hpp
   InsertInterpolationScriData.hpp
```
Alphabetize.

wthrowe commented on src/Evolution/Systems/Cce/Actions/ReceiveGhWorldtubeData.hpp:
```diff
@@ -45,7 +45,7 @@ namespace Actions {
  *   - `Tags::GhInterfaceManager`
  */
-template <typename EvolutionComponent>
+template <typename EvolutionComponent, bool DuringSelfStart>
```
Maybe merge this with the next commit where the template parameter is used?

kidder committed:
Split off LinkedMessageId into its own file

Also, add a custom less comparator for LinkedMessageId.

markscheel commented on src/ApparentHorizons/ComputeHorizonVolumeQuantities.hpp:
```diff
@@ -68,12 +68,6 @@ struct ComputeHorizonVolumeQuantities
       const InverseJacobian<DataVector, 3, Frame::ElementLogical, TargetFrame>&
           inverse_jacobian);
 
-  using allowed_src_tags = tmpl::list<
```
Why is CCE using ComputeHorizonVolumeQuantities?   CCE should use a different class that conforms to `intrp::protocols::ComputeVarsToInterpolate`.   (Similarly, interpolation onto an excision bdry uses a different class as well).  Presumably whatever CCE needs should be simpler than ComputeHorizonVolumeQuantities because CCE (I think) needs only metric and 1st derivs, not 2nd derivs.

If it so happens that CCE needs (a subset of) `ComputeHorizonVolumeQuantities` as-is, then I can see the case for not writing another class with a subset of the functionality.  But in that case, `ComputeHorizonVolumeQuantities` should be renamed.  (and I am fine with removing `allowed_src_tags` to allow passing in a Variables with extra tags).

(Also, note that by removing `allowed_src_tags`, the user might assume that any Tags in the SrcVariables that are not in `required_src_tags` remain unused.  But that is not true. I don't know how to get around this other than better dox.)

kidder commented:
@wthrowe @knelli2 @markscheel    I have rebased and squashed making the following changes:

- Removed the commit introducing the TimeAndPrevious struct and instead use the almost identical LinkedMessageId template
- Added a commit that factored out LinkedMessageId form LinkedMessageQueue
- Anywhere that used a TimeAndPrevious now uses a `LinkedMessageId<double>`
- Dropped the commit removing `allowed_src_tags` from `ComputeVolumeHorizonQuantities`
- Added the missing protocol includes
- Alphabetized in `Cce/Actions/CMakeLists.txt`
- Merged two commits as suggested by @wthrowe 

I agree with @markscheel that CCE should probably not reuse `ComputeVolumeHorizonQuantites` (and any other Horizon related interpolation target classes), but think that is beyond the scope of this PR

kidder commented:
Labelling as in progress until someone runs CCE and verifies nothing is broken

knelli2 reviewed (approved):
@kidder Yes, as we use the interpolation framework for more things other than horizon finding, we'll need more/different ComputeVarsToInterpolate classes. But I agree, that's beyond this PR

markscheel reviewed (approved):
Yes, agreed that in general we should have multiple ComputeVarsToInterpolate classes.

However, if it turns out that two ComputeVarsToInterpolate classes duplicate the same code, then we might want to consider using only one of them (and renaming it to a more general name), to avoid code duplication.

Sizheng-Ma commented:
I've checked that the CCE results are consistent with develop.
